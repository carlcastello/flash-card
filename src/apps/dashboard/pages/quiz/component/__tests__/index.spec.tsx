import React from "react";
import { shallow } from "enzyme";

import { Quiz } from "../index";
import toJson from "enzyme-to-json";
import { Redirect } from "react-router-dom";
import LoadingScreen from "../../../../../commons/components/loading-screen";
import { Grid } from "@material-ui/core";

describe("Quiz", () => {
  let id: string = "hello-world";
  
  let fetchQuizFunction: any;
  let createQuizSummaryFunction: any;
  let saveQuizSummaryFunction: any;
  let createQuestionFunction: any;
  let saveQuestionFunction: any;
  let deleteQuestionFunction: any;

  let wrapper: any;

  beforeEach(() => {

    fetchQuizFunction = jest.fn();
    createQuizSummaryFunction = jest.fn();
    saveQuizSummaryFunction = jest.fn();
    createQuestionFunction = jest.fn();
    saveQuestionFunction = jest.fn();
    deleteQuestionFunction = jest.fn();

    wrapper = shallow(
      <Quiz
        isFullPageLoading={false}
        creatingQuizSummary={false}
        savingQuizSummary={false}
        isCreatingQuestion={false}
        isSavingQuestion={false}
        isDeletingQuestion={false}
      
        requiredData={[]}
        quiz={{
          id,
          quizSummary: {
            title: "This is the title",
            description: "Description"
          },
          quizQuestions: [{
            id: "question-id",
            question: "this is a question",
            subQuestion: "this is sub question",
            answer: "string",
          }]
        }}   
        fetchQuiz={fetchQuizFunction}
        createQuizSummary={createQuizSummaryFunction}
        saveQuizSummary={saveQuizSummaryFunction}
        createQuestion={createQuestionFunction}
        saveQuestion={saveQuestionFunction}
        deleteQuestion={deleteQuestionFunction}    
        classes={{
          gridContainer: "grid-container",
        }}
        history={{ push: jest.fn() } as any}
        location={{} as any}
        match={{
          params: {
            quizId: id
          }
        } as any}
      />
    )
  });

  describe("The Render Function", () => {
    it("renders the redirect", () => {
      wrapper.setProps({ quiz: { id }, match: { params: {} } });
      expect(toJson(wrapper)).toMatchSnapshot();

      expect(wrapper.find(Redirect).props().to)
        .toEqual("/dashboard/quiz/hello-world");
    });

    it("renders the loading page when page is loading", () => {
      wrapper.setProps({ 
        isFullPageLoading: true,
      });
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(LoadingScreen)).toHaveLength(1);
      expect(wrapper.find(Grid)).toHaveLength(0);
    });

    it("renders the loading page when creatingQuizSummary", () => {
      wrapper.setProps({ 
        creatingQuizSummary: true,
      });
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(LoadingScreen)).toHaveLength(1);
      expect(wrapper.find(Grid)).toHaveLength(0);
    });

    it("renders just the quizSummary questionaire", () => {
      wrapper.setProps({
         quiz: {},
         match: { params: { quizId: null} }
      });
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(Grid)).toHaveLength(2);
    });

    it("renders the full quiz", () => {
      wrapper.setProps({
         quiz: { id },
         match: { params: { quizId: id} }
      });
      expect(toJson(wrapper)).toMatchSnapshot();
      expect(wrapper.find(Grid)).toHaveLength(3);
    });
  });
});