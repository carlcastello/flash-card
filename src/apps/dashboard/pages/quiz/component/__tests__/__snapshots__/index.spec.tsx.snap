// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Quiz The Render Function renders just the quizSummary questionaire 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="grid-container"
  container={true}
  direction="row"
  justify="center"
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    sm={10}
  >
    <WithStyles(QuizSummary)
      isLoading={false}
      onCreateQuizSummary={[MockFunction]}
      onSaveQuizSummary={[MockFunction]}
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Quiz The Render Function renders the full quiz 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="grid-container"
  container={true}
  direction="row"
  justify="center"
  spacing={2}
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    sm={10}
  >
    <WithStyles(QuizSummary)
      isLoading={false}
      onCreateQuizSummary={[MockFunction]}
      onSaveQuizSummary={[MockFunction]}
      quizId="hello-world"
    />
  </WithStyles(ForwardRef(Grid))>
  <WithStyles(ForwardRef(Grid))
    item={true}
    sm={10}
  >
    <WithStyles(QuizQuestion)
      createQuestion={[MockFunction]}
      deleteQuestion={[MockFunction]}
      isCreatingQuestion={false}
      isDeletingQuestion={false}
      isSavingQuestion={false}
      push={[MockFunction]}
      quizId="hello-world"
      quizQuestions={Array []}
      saveQuestion={[MockFunction]}
    />
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;

exports[`Quiz The Render Function renders the loading page when creatingQuizSummary 1`] = `
<WithStyles(LoadingScreen)>
  Saving Created Quiz...
</WithStyles(LoadingScreen)>
`;

exports[`Quiz The Render Function renders the loading page when page is loading 1`] = `
<WithStyles(LoadingScreen)>
  Loading Quiz Data...
</WithStyles(LoadingScreen)>
`;

exports[`Quiz The Render Function renders the redirect 1`] = `
<Redirect
  to="/dashboard/quiz/hello-world"
/>
`;
